File: usr/local/bin/frognet_monitor_cs/Parsers.cs
/***************************************************************
 *  Copyright (C) 2016-2026 Fawcett Innovations LLC            *
 *                                                             *
 *  SPDX-License-Identifier: GPL-2.0-only                      *
 *                                                             *
 *  This program is free software; you can redistribute it     *
 *  and/or modify it under the terms of the GNU General Public *
 *  License as published by the Free Software Foundation;      *
 *  version 2 of the License, and no other version.            *
 *                                                             *
 *  This program is distributed in the hope that it will be    *
 *  useful, but WITHOUT ANY WARRANTY; without even the implied *
 *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR    *
 *  PURPOSE.  See the GNU General Public License for details.  *
 *                                                             *
 *  See COPYRIGHT and LICENSE at the root of this tree.        *
 **************************************************************/
namespace FrogNet.Monitor;

/// <summary>
/// The three pure parsers. No I/O, no sockets, no clock.
///
/// These are exercised by ../frognet_monitor_shared/parser_vectors.json — the
/// same file the Python and C++ implementations read — so agreement between the
/// ports is a test result rather than a claim in a README.
/// </summary>
public static class Parsers
{
    /// <summary>Identity as returned by frognet_echo.php / getFrogNet.bash.</summary>
    public readonly record struct Identity(string Domain, string Eth0, string Wlan0, string Wlan1);

    /// <summary>A FrogNet node as it appears in /etc/hosts or getHosts.php.</summary>
    public readonly record struct HostEntry(string Ip, string Name);

    /// <summary>
    /// Parse one CSV line: <c>fqdn,eth0IP,wlan0IP,wlan1IP</c>.
    /// Returns null when identity failed.
    ///
    /// [IDENTITY_FAILS_LOUD_V1] Four fields with a non-empty NAME, or it failed.
    /// getFrogNet.bash exits non-zero and prints nothing when a node cannot state
    /// its identity, so the echo body is empty and there is nothing to guess. No
    /// fallback name is reconstructed: every node answers to the hostname
    /// "FrogNetHost", so a name that is not the domain is not an identity, and a
    /// client that accepts one believes every node is the same node.
    ///
    /// An empty INTERFACE field is accurate data — a node with no carrier on eth0
    /// genuinely has no eth0 address.
    /// </summary>
    public static Identity? ParseEchoLine(string? body)
    {
        if (string.IsNullOrWhiteSpace(body)) return null;

        // First non-empty line; the body may carry a trailing newline.
        var line = body.Trim();
        var nl = line.IndexOf('\n');
        if (nl >= 0) line = line[..nl].Trim();
        if (line.Length == 0) return null;

        // NOTE: plain Split, NOT StringSplitOptions.RemoveEmptyEntries.
        //
        // This is the single most important line in the C# port. With
        // RemoveEmptyEntries, "Seattle2,,10.120.120.1,10.160.160.47" becomes
        // three fields and every subsequent value shifts left — wlan0 would hold
        // the wlan1 address. The result is well-formed, plausible, and wrong,
        // and nothing downstream can detect it. The shared vectors exist
        // largely to pin this behaviour.
        var f = line.Split(',');
        if (f.Length != 4) return null;

        for (var i = 0; i < f.Length; i++) f[i] = f[i].Trim();
        if (f[0].Length == 0) return null;   // empty NAME is a failure

        return new Identity(f[0], f[1], f[2], f[3]);
    }

    /// <summary>
    /// Parse /etc/hosts content into FrogNet nodes.
    ///
    /// 10/8 only; 10.253.0.0/16 (tunnel transit) and 10.254.0.0/16 (chorus
    /// virtual) excluded; a <c>FrogNetHost.&lt;name&gt;</c> alias required, and
    /// the name is the part after it. Every node's hostname is FrogNetHost, so
    /// the alias suffix is the only thing on the line identifying which node it
    /// describes.
    /// </summary>
    public static List<HostEntry> ParseEtcHosts(string? content)
    {
        var result = new List<HostEntry>();
        if (string.IsNullOrEmpty(content)) return result;

        const string marker = "FrogNetHost.";
        foreach (var raw in content.Split('\n'))
        {
            var line = raw.Trim();
            if (line.Length == 0 || line[0] == '#') continue;

            var parts = line.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length < 2) continue;

            var ip = parts[0];
            if (!ip.StartsWith("10.", StringComparison.Ordinal)) continue;
            if (ip.StartsWith("10.253.", StringComparison.Ordinal)) continue;
            if (ip.StartsWith("10.254.", StringComparison.Ordinal)) continue;

            for (var i = 1; i < parts.Length; i++)
            {
                var at = parts[i].IndexOf(marker, StringComparison.Ordinal);
                if (at < 0) continue;
                var name = parts[i][(at + marker.Length)..];
                if (name.Length > 0) result.Add(new HostEntry(ip, name));
                break;
            }
        }
        return result;
    }

    /// <summary>
    /// Own address to the .1 that can state this pond's identity.
    /// Returns "" when the address is not a basis for one: 10.253/16 and
    /// 10.254/16 are carried by a node but do not identify it.
    /// </summary>
    public static string DotOneOf(string? ip)
    {
        if (string.IsNullOrWhiteSpace(ip)) return "";
        var s = ip.Trim();
        if (!s.StartsWith("10.", StringComparison.Ordinal)) return "";
        if (s.StartsWith("10.253.", StringComparison.Ordinal)) return "";
        if (s.StartsWith("10.254.", StringComparison.Ordinal)) return "";

        var o = s.Split('.');
        if (o.Length != 4) return "";
        foreach (var part in o)
        {
            if (part.Length == 0 || part.Length > 3) return "";
            foreach (var c in part) if (c < '0' || c > '9') return "";
            if (int.Parse(part) > 255) return "";
        }
        return $"{o[0]}.{o[1]}.{o[2]}.1";
    }
}

File: usr/local/bin/frognet_monitor_cpp/src/parsers.cpp
/***************************************************************
 *  Copyright (C) 2016-2026 Fawcett Innovations LLC            *
 *                                                             *
 *  SPDX-License-Identifier: GPL-2.0-only                      *
 *                                                             *
 *  This program is free software; you can redistribute it     *
 *  and/or modify it under the terms of the GNU General Public *
 *  License as published by the Free Software Foundation;      *
 *  version 2 of the License, and no other version.            *
 *                                                             *
 *  This program is distributed in the hope that it will be    *
 *  useful, but WITHOUT ANY WARRANTY; without even the implied *
 *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR    *
 *  PURPOSE.  See the GNU General Public License for details.  *
 *                                                             *
 *  See COPYRIGHT and LICENSE at the root of this tree.        *
 **************************************************************/
// parsers.cpp — the three pure functions. No I/O, no sockets, no clock.
//
// These are exercised by ../frognet_monitor_shared/parser_vectors.json, the same
// file the Python and C# implementations read, so agreement between the ports is
// a test result rather than a claim.
#include "frognet.h"

#include <algorithm>
#include <cctype>
#include <sstream>

namespace frognet {
namespace {

std::string trim(const std::string& s) {
  size_t a = 0, b = s.size();
  while (a < b && std::isspace(static_cast<unsigned char>(s[a]))) ++a;
  while (b > a && std::isspace(static_cast<unsigned char>(s[b - 1]))) --b;
  return s.substr(a, b - a);
}

// Split on a delimiter WITHOUT collapsing adjacent ones and WITHOUT dropping
// empties. This is the whole reason the shared vectors exist: strtok collapses
// runs of delimiters, and C#'s Split with RemoveEmptyEntries drops empties, so
// "Seattle2,,10.120.120.1,10.160.160.47" silently becomes a 3-field line whose
// wlan0 holds the wlan1 address. Wrong, and it looks right.
std::vector<std::string> splitKeepEmpty(const std::string& s, char delim) {
  std::vector<std::string> out;
  std::string cur;
  for (char c : s) {
    if (c == delim) { out.push_back(cur); cur.clear(); }
    else cur += c;
  }
  out.push_back(cur);
  return out;
}

std::vector<std::string> splitWhitespace(const std::string& s) {
  std::vector<std::string> out;
  std::istringstream is(s);
  std::string tok;
  while (is >> tok) out.push_back(tok);
  return out;
}

bool startsWith(const std::string& s, const std::string& p) {
  return s.size() >= p.size() && s.compare(0, p.size(), p) == 0;
}

}  // namespace

Identity parseEchoLine(const std::string& body) {
  Identity id;

  // The body may carry a trailing newline; take the first non-empty line and
  // ignore anything after it.
  std::string line = trim(body);
  size_t nl = line.find('\n');
  if (nl != std::string::npos) line = trim(line.substr(0, nl));

  // [IDENTITY_FAILS_LOUD_V1] Empty body means the node could not state its
  // identity. getFrogNet.bash exits non-zero and prints nothing, so
  // frognet_echo.php returns an empty body. There is no fallback name to
  // reconstruct and nothing to guess: it failed.
  if (line.empty()) return id;

  std::vector<std::string> f = splitKeepEmpty(line, ',');
  if (f.size() != 4) return id;      // exactly four fields, or it failed

  for (auto& x : f) x = trim(x);

  // An empty INTERFACE field is accurate data — Seattle2 has no carrier on eth0
  // and reporting that honestly is correct. An empty NAME is a failure, because
  // every node answers to the hostname "FrogNetHost" and only the domain
  // identifies which node this is.
  if (f[0].empty()) return id;

  id.ok = true;
  id.domain = f[0];
  id.eth0   = f[1];
  id.wlan0  = f[2];
  id.wlan1  = f[3];
  return id;
}

std::vector<HostEntry> parseEtcHosts(const std::string& content) {
  std::vector<HostEntry> out;
  std::istringstream is(content);
  std::string raw;
  while (std::getline(is, raw)) {
    std::string line = trim(raw);
    if (line.empty() || line[0] == '#') continue;

    std::vector<std::string> parts = splitWhitespace(line);
    if (parts.size() < 2) continue;

    const std::string& ip = parts[0];
    // 10/8 only. 10.253/16 is tunnel transit and 10.254/16 is chorus virtual —
    // a node carries those addresses but is not reachable as a node at them.
    if (!startsWith(ip, "10.")) continue;
    if (startsWith(ip, "10.253.") || startsWith(ip, "10.254.")) continue;

    // A FrogNetHost.<name> alias is required, and the name is what follows it.
    // Every node's hostname is FrogNetHost, so the suffix is the only thing on
    // the line that says which node it is.
    static const std::string kMarker = "FrogNetHost.";
    for (size_t k = 1; k < parts.size(); ++k) {
      size_t at = parts[k].find(kMarker);
      if (at == std::string::npos) continue;
      std::string name = parts[k].substr(at + kMarker.size());
      if (!name.empty()) out.push_back(HostEntry{ip, name});
      break;
    }
  }
  return out;
}

std::string dotOneOf(const std::string& ip) {
  std::string s = trim(ip);
  if (!startsWith(s, "10.")) return "";
  if (startsWith(s, "10.253.") || startsWith(s, "10.254.")) return "";

  std::vector<std::string> o = splitKeepEmpty(s, '.');
  if (o.size() != 4) return "";
  for (const auto& part : o) {
    if (part.empty() || part.size() > 3) return "";
    for (char c : part) if (!std::isdigit(static_cast<unsigned char>(c))) return "";
    int v = std::atoi(part.c_str());
    if (v < 0 || v > 255) return "";
  }
  return o[0] + "." + o[1] + "." + o[2] + ".1";
}

}  // namespace frognet
